Спринт 16.

Kris-found opened the pull request:
- Переписан код с использованием шаблона MVVM
- Реорганизована структура приложения

Kris-found committed:
Спринт 16.
- Переписан код с использованием шаблона MVVM
- Реорганизована структура приложения

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/search/data/network/RetrofitNetworkClientImpl.kt:
```diff
@@ -0,0 +1,58 @@
+        .addConverterFactory(GsonConverterFactory.create())
+        .build()
+
+    private val iTunesService = retrofit.create(ITunesSearchAPI::class.java)
```
Рекомендация:
Это зависимость, которую стоит создавать в `Creator` и передавать сюда через конструктор класса.

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/search/ui/TracksSearchViewModel.kt:
```diff
@@ -0,0 +1,137 @@
+        handler.postDelayed(searchRunnable, SEARCH_DEBOUNCE_DELAY_IN_SECONDS)
+    }
+
+    fun clickDebounce(track: Tracks): Boolean {
```
Обязательно:
Публичные методы не должны возвращать ни каких значений обратно во активити/фрагмент

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/search/ui/TracksSearchViewModel.kt:
```diff
@@ -0,0 +1,137 @@
+
+    private val handler = Handler(Looper.getMainLooper())
+
+    private val searchStateLiveData = MutableLiveData<SearchState>()
```
Обязательно:
Не должно быть несколько разных стейтов. Нужно сформировать единый стейт экрана объединив `SearchState` и `HistoryState` в один `SearchScreenState`

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/player/ui/AudioPlayerViewModel.kt:
```diff
@@ -0,0 +1,92 @@
+    private val playerInteractor: AudioPlayerInteractor,
+) : ViewModel() {
+
+    private val playerStateLiveData = MutableLiveData<PlayerState>()
```
Обязательно:
Здесь аналогично, нужно сформировать единый стейт из `PlayerState` и времени воспроизведения.
Для этого хорошо подойдёт `data class` и/или `sealed class`

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/search/domain/model/Tracks.kt:
```diff
@@ -1,4 +1,4 @@
-package com.practicum.playlistmaker.domain.model
+package com.practicum.playlistmaker.search.domain.model
 
 import android.os.Parcelable
```
Обязательно:
Классы в пакете `domain` не должны использовать классы из Android SDK.
Здесь можно заменить на `Serializable` или выделить презентационную модель и добавить маппинг между `presentation` и `domain` слоями.

Kris-found committed:
Спринт 16(правки)
- Исправлен баг: при окончании воспроизведения кнопка и счетчик не менялись
- у ViewModel нет возвращаемых значений у публичных методов
- создан единый стейт во ViewModel у экранов: поиск и плейер
- заменено использование Parcelable в модели Tracks пакет domain на Serializable
